plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+                for digest_name in declarative_artifact.artifact.DIGEST_FIELDS:
+                    digest_value = getattr(declarative_artifact.artifact, digest_name)
+                    if digest_value:
+                        key = {digest_name: digest_value}
+                        one_artifact_q &= Q(**key)
+                all_artifacts_q |= one_artifact_q
+
+        for artifact in Artifact.objects.filter(all_artifacts_q):
+            for content in declarative_content:
+                if content is None:
+                    continue
+                for declarative_artifact in content.d_artifacts:
+                    for digest_name in artifact.DIGEST_FIELDS:
+                        digest_value = getattr(declarative_artifact.artifact, digest_name)
+                        if digest_value and digest_value == getattr(artifact, digest_name):
+                            declarative_artifact.artifact = artifact
+        for content in declarative_content:
+            if content is None:
+                continue
+            await out_q.put(content)
+        declarative_content = []
+        if shutdown:
+            break
+    await out_q.put(None)
+
+
+class artifact_downloader:
+    """
+    An object containing a Stages API stage to download :class:`~pulpcore.plugin.models.Artifact`
+    file, but don't save the :class:`~pulpcore.plugin.models.Artifact` in the db.
+
+    The actual stage is the :meth:`~pulpcore.plugin.stages.artifact_downloader.stage`
+    which can be used as follows:
+
+    >>> artifact_downloader(max_concurrent_downloads=42).stage  # This is the real stage
+
+    This stage downloads the file for any :class:`~pulpcore.plugin.models.Artifact` objects missing
+    files and creates a new :class:`~pulpcore.plugin.models.Artifact` object from the downloaded
+    file and its digest data. The new :class:`~pulpcore.plugin.models.Artifact` is not saved but
+    added to the :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object, replacing the likely
+    incomplete :class:`~pulpcore.plugin.models.Artifact`.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage creates a ProgressBar named 'Downloading Artifacts' that counts the number of
+    downloads completed. Since it's a stream the total count isn't known until it's finished.
+
+    This stage drains all available items from `in_q` and starts as many downloaders as possible.
+
+    Args:
+        max_concurrent_downloads (int): The maximum number of concurrent downloads this stage will
+            run. Default is 100.
+
+    Returns:
+        An object containing the artifact_downloader stage to be included in a pipeline.
+    """
+
+    def __init__(self, max_concurrent_downloads=100):
+        self.max_concurrent_downloads = max_concurrent_downloads
+
+    async def stage(self, in_q, out_q):
```

Comment:
This method contains multiple code blocks each doing something complex.  Recommend decomposing into smaller methods each doing *one* thing.  Better functional decomposition will be more readable, testable and maintainable.

Reply:
I agree, but I can't do that now. See reasons [here](https://github.com/pulp/pulp/pull/3559#discussion_r206947935).